clang/lib/Sema/SemaDecl.cpp

```diff
@@ -3922,6 +3922,42 @@ bool Sema::MergeFunctionDecl(FunctionDecl *New, NamedDecl *&OldD, Scope *S,
     return true;
   }
 
+  const auto OldFX = Old->getFunctionEffects();
+  const auto NewFX = New->getFunctionEffects();
+  if (OldFX != NewFX) {
+    const auto Diffs = FunctionEffectSet::differences(OldFX, NewFX);
+    for (const auto &Item : Diffs) {
+      const FunctionEffect *Effect = Item.first;
+      const bool Adding = Item.second;
+      if (Effect->diagnoseRedeclaration(Adding, *Old, OldFX, *New, NewFX)) {
+        Diag(New->getLocation(),
+             diag::warn_mismatched_func_effect_redeclaration)
+            << Effect->name();
+        Diag(Old->getLocation(), diag::note_previous_declaration);
+      }
+    }
+
+    const auto MergedFX = OldFX | NewFX;
+
+    // Having diagnosed any problems, prevent further errors by applying the
+    // merged set of effects to both declarations.
+    auto applyMergedFX = [&](FunctionDecl *FD) {
+      const auto *FPT = FD->getType()->getAs<FunctionProtoType>();
+      FunctionProtoType::ExtProtoInfo EPI = FPT->getExtProtoInfo();
+      EPI.FunctionEffects = MergedFX;
+      QualType ModQT = Context.getFunctionType(FD->getReturnType(),
+                                               FPT->getParamTypes(), EPI);
+
+      FD->setType(ModQT);
+    };
+
+    applyMergedFX(Old);
+    applyMergedFX(New);
+
+    OldQType = Old->getType();
```

Comment:
In prototyping the feature, it was really helpful to be able to redeclare things like `pthread_self()` as `nolock`. This was also envisioned as a viable bootstrapping technique.

Maybe, however, this is an undesirable hack. A workaround would be to create wrapper functions that are declared safe but call the unsafe function with diagnostics disabled. To facilitate this a bit more cleanly, we could add the concept of a "leaf" function (as with TCB). `nolock_leaf` or `nolock(leaf)` (?) could mean "tell my callers I am `nolock`, but don't verify me."

Reply:
> In prototyping the feature, it was really helpful to be able to redeclare things like `pthread_self()` as `nolock`.

In that case, when determining whether a function should be treated as `nolock` during the analysis phase, one option would be to check all of its declarations and consider a function to be `nolock` if at least one of its declaration is `nolock`.

The problem I see w/ doing something here is that we don’t modify the `Old` declaration pretty much ever, from what I remember, so I’m somewhat wary of doing so breaking things in unforeseen ways, because the rest of the code base can currently assume that we don’t just go around changing function decls all the time—that said, I’m not *sure* doing so would break something, just a bit worried it might.

Reply:
> A workaround would be to create wrapper functions that are declared safe but call the unsafe function with diagnostics disabled.

I could see there being situations where you might want to be able to do something like that (i.e. declare a function as safe even if the compiler might think it’s unsafe because it calls unsafe functions), but I wonder if you couldn’t just disable the diagnostics locally in that function (or even just parts thereof) only using a `#pragma`.

Reply:
> we don’t modify the `Old` declaration pretty much ever, from what I remember

That said, if you can find a place in the entire function merging machinery where we *do* modify  `Old`, then it might be fine.

Reply:
I personally feel like the macro solution is ‘good enough’, if you will; I don’t think `CheckEquivalentExceptionSpec` modifies `Old`, so I’m not sure why that comment says is does—maybe it did at the time (the comment is from 2016, and, funnily enough, was also added in the same commit that introduced that call to `isFunctionConversion()` where the arguments seem transposed).

There is one place where we call `setDeletedAsWritten(false)` on the old decl in a caller of `MergeFunctionDecl()`, but that one has a FIXME next to it, so...

I’ll ask if anyone’s aware of any specific reasons why or ways in which we should avoid modifying `Old` here, and if not, then I’ll see if I can make the pointer `const`, because that would communicate the intent more clearly that it’s really not supposed to be modified.

Reply:
@dougsonos So, update, I just talked to Aaron about this, and generally speaking, it’s better to avoid modifying the `Old` declaration (even in cases where it wouldn’t change the behaviour of the program) as that may be a source of possible confusion for e.g. tools that are consuming the AST but don’t know about this.

I think I’ve mentioned this already, but an alternative that comes to mind would be to just look at all declarations of a function (starting at the most recent one, because that one will most likely have the attribute on it) and consider a function to be effectively `nolock`/`noalloc` if any of them are `nolock`/`noalloc`. Iirc, there was some code somewhere that is only checking the canonical decl, so that would have to be updated as a result.

Reply:
> it’s better to avoid modifying the `Old` declaration

OK, I agree that that will make more sense.

> look at all declarations of a function (starting at the most recent one, because that one will most likely have the attribute on it) 

There are quite a number of places that ask this question but I'll work through this. This is an interesting example:
```
void foo() {
	auto* x = new int;
}
void foo() [[clang::nolock]];
void bar() [[clang::nolock]] {
	foo();
}
```
I think that `foo()` ought to get verified because by the time verification happens, the redeclaration has been seen.

I'm having difficulty testing this though, because now that the `Old` declaration is untouched, `MergeFunctionDecl` is unhappy that `New` has a different canonical type. Would it make sense to add another check near the end like this one?
```
  // Check if the function types are compatible when pointer size address
  // spaces are ignored.
  if (Context.hasSameFunctionTypeIgnoringPtrSizes(OldQType, NewQType))
    return false;
```

Reply:
I've made `Sema::MergeFunctionDecl` use `OldQTypeForComparison` earlier so as to ignore effect differences in redeclarations, and `FunctionDecl::getFunctionEffects()` now collects and returns the union of effects present in all redeclarations.

Reply:
Sounds good to me.